File: examples/KelvinTransformation/H-formulation/Fig_3.m
clear all;
close all;

set(gcf,'Units','Pixels','Position',[100,30,1200,500],'PaperSize',[1200/96*2.54,500/96*2.54]);
for n = 1:2
	a(n) = axes('Units','Pixels','Position',[155+(n-1)*600,135,425,345],'FontName','Times New Roman','FontSize',36);
	hold on;	box on;	grid on; grid minor;
	xlabel('{\it x} (m)');
	ylabel('{\it H_z} (A/m)');
	set(gca,'XLim',[-1.1, 1.1], 'XTick', [-1.0:1.00:1.0]);
end

theta = linspace(0,2*pi,101);

for n = 1:2
	axes(a(n))
		switch n
		case 1
			load('3D_dipole_with_Kelvin.mat');
			x = xx(111,:);
			Hz = Hz_analytical(111,:);
			Hz(find(abs(x)>=0.5)) = - Hz(find(abs(x)>=0.5));
			plot(x, Hz, 'k-', 'LineWidth', 1.0);

			load('3D_dipole.mat');
			x = xx(111,:);
			Hz = Hz(111,:);
			plot(x, Hz, 'r-');

			load('3D_dipole_with_Kelvin.mat');
			x = xx(111,:);
			Hz = Hz(111,:);
			plot(x, Hz, 'c--', 'LineWidth', 1.5);

			set(gca,'YLim',[-1.1, 0.0]);
			text(0.75,-1.4,'(a)','FontName','Times New Roman','FontSize', 36);
			h = legend({'Analytical', 'Finite domain' 'Kelvin trans.'},'location','best','box','off', 'FontName','Times New Roman','FontSize', 28);
		case 2
			load('3D_quadrupole_with_Kelvin.mat');
			x = xx(111,:);
			Hz = Hz_analytical(111,:);
			Hz(find(abs(x)>=0.5)) = Hz(find(abs(x)>=0.5));
			plot(x, Hz, 'k-', 'LineWidth', 1.0);

			load('3D_quadrupole.mat');
			x = xx(111,:);
			Hz = Hz(111,:);
			plot(x, Hz, 'r-');

			load('3D_quadrupole_with_Kelvin.mat');
			x = xx(111,:);
			Hz = Hz(111,:);
			plot(x, Hz, 'c--', 'LineWidth', 1.5);

			set(gca,'YLim',[-0.60, 0.60]);
			text(0.75,-0.93,'(b)','FontName','Times New Roman','FontSize', 36);
			h = legend({'Analytical', 'Finite domain' 'Kelvin trans.'},'location','southeast','box','off', 'FontName','Times New Roman','FontSize', 28);
		end
end

exportgraphics(gcf, sprintf('%s.pdf', mfilename));

File: examples/kelvin_transformation/H-formulation/Fig_3.py
"""Fig_3: 1x2 panel line plot comparing Hz along x-axis.

Ported from Fig_3.m (MATLAB). Loads .mat data files produced by
3D_dipole.py, 3D_dipole_with_Kelvin.py,
3D_quadrupole.py, 3D_quadrupole_with_Kelvin.py.

Panels:
  (a) Dipole Hz(x) comparison: analytical vs finite domain vs Kelvin
  (b) Quadrupole Hz(x) comparison: analytical vs finite domain vs Kelvin
"""
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import loadmat


def main():
    fig, axes = plt.subplots(1, 2, figsize=(12.5, 5.2))
    row_idx = 110  # corresponds to MATLAB index 111 (0-based)

    # (a) Dipole
    ax = axes[0]

    data_k = loadmat("3D_dipole_with_Kelvin.mat")
    x = data_k["xx"][row_idx, :]
    Hz_ana = data_k["Hz_analytical"][row_idx, :].copy()
    Hz_ana[np.abs(x) >= 0.5] = -Hz_ana[np.abs(x) >= 0.5]
    ax.plot(x, Hz_ana, "k-", linewidth=1.0, label="Analytical")

    data_f = loadmat("3D_dipole.mat")
    x_f = data_f["xx"][row_idx, :]
    Hz_f = data_f["Hz"][row_idx, :]
    ax.plot(x_f, Hz_f, "r-", label="Finite domain")

    Hz_kelvin = data_k["Hz"][row_idx, :]
    ax.plot(x, Hz_kelvin, "c--", linewidth=1.5, label="Kelvin trans.")

    ax.set_xlim(-1.1, 1.1)
    ax.set_ylim(-1.1, 0.0)
    ax.set_xlabel("x (m)")
    ax.set_ylabel("Hz (A/m)")
    ax.text(0.75, -1.35, "(a)", fontsize=14)
    ax.legend(loc="best", frameon=False, fontsize=11)
    ax.grid(True)

    # (b) Quadrupole
    ax = axes[1]

    data_k = loadmat("3D_quadrupole_with_Kelvin.mat")
    x = data_k["xx"][row_idx, :]
    Hz_ana = data_k["Hz_analytical"][row_idx, :]
    ax.plot(x, Hz_ana, "k-", linewidth=1.0, label="Analytical")

    data_f = loadmat("3D_quadrupole.mat")
    x_f = data_f["xx"][row_idx, :]
    Hz_f = data_f["Hz"][row_idx, :]
    ax.plot(x_f, Hz_f, "r-", label="Finite domain")

    Hz_kelvin = data_k["Hz"][row_idx, :]
    ax.plot(x, Hz_kelvin, "c--", linewidth=1.5, label="Kelvin trans.")

    ax.set_xlim(-1.1, 1.1)
    ax.set_ylim(-0.60, 0.60)
    ax.set_xlabel("x (m)")
    ax.set_ylabel("Hz (A/m)")
    ax.text(0.75, -0.90, "(b)", fontsize=14)
    ax.legend(loc="lower right", frameon=False, fontsize=11)
    ax.grid(True)

    plt.tight_layout()
    plt.savefig("Fig_3.pdf")
    plt.show()


if __name__ == "__main__":
    main()
